src/useBlocklyWorkspace.js

```diff
@@ -0,0 +1,130 @@
+  try {
+    Blockly.Xml.domToWorkspace(Blockly.Xml.textToDom(xml), workspace);
+    return true;
+  } catch (e) {
+    if (onImportXmlError) {
+      onImportXmlError(e);
+    }
+    return false;
+  }
+}
+
+const useBlocklyWorkspace = ({
+  ref,
+  initialXml,
+  toolboxConfiguration,
+  workspaceConfiguration,
+  onWorkspaceChange,
+  onImportXmlError,
+}) => {
+  const [workspace, setWorkspace] = React.useState(null);
+  const [xml, setXml] = React.useState(initialXml);
+  const [didInitialImport, setDidInitialImport] = React.useState(false);
+  const [didHandleNewWorkspace, setDidHandleNewWorkspace] = React.useState(false);
+
+  // we explicitly don't want to recreate the workspace when the configuration changes
+  // so, we'll keep it in a ref and update as necessary in an effect hook
+  const workspaceConfigurationRef = React.useRef(workspaceConfiguration);
+  React.useEffect(() => {
+    workspaceConfigurationRef.current = workspaceConfiguration;
+  }, [workspaceConfiguration]);
+
+  const toolboxConfigurationRef = React.useRef(toolboxConfiguration);
+  React.useEffect(() => {
+    toolboxConfigurationRef.current = toolboxConfiguration;
+    if (toolboxConfiguration && workspace) {
+      workspace.updateToolbox(toolboxConfiguration);
+    }
+  }, [toolboxConfiguration, workspace]);
+
+  const handleWorkspaceChanged = React.useCallback(
+    (newWorkspace) => {
+      if (onWorkspaceChange) {
+        onWorkspaceChange(newWorkspace);
+      }
+    },
+    [onWorkspaceChange]
+  );
+
+  // Workspace creation
+  React.useEffect(() => {
+    const newWorkspace = Blockly.inject(ref.current, {
+      ...workspaceConfigurationRef.current,
+      toolbox: toolboxConfigurationRef.current,
+    });
+    setWorkspace(newWorkspace);
+    setDidInitialImport(false); // force a re-import if we recreate the workspace
+    setDidHandleNewWorkspace(false); // Singal that a workspace change event needs to be sent.
+
+    // Dispose of the workspace when our div ref goes away (Equivalent to didComponentUnmount)
+    return () => {
+      newWorkspace.dispose();
+    };
+  }, [ref]);
+
+  // Send a workspace change event when the workspace is created
+  React.useEffect(() => {
+    if (workspace && !didHandleNewWorkspace) {
+      handleWorkspaceChanged(workspace);
+    }
+  }, [handleWorkspaceChanged, didHandleNewWorkspace, workspace]);
+
+  // Workspace change listener
+  React.useEffect(() => {
+    if (workspace == null) {
+      return undefined;
+    }
+    
+    const listener = () => { handleWorkspaceChanged(workspace); };
+    workspace.addChangeListener(listener);
+    return () => {
+      workspace.removeChangeListener(listener);
+    };
+  }, [workspace, handleWorkspaceChanged]);
+
+  // xmlDidChange callback
+  React.useEffect(() => {
+    if (workspace == null) {
+      return undefined;
+    }
+
+    const [callback, cancel] = debounce(() => {
+      const newXml = Blockly.Xml.domToText(
```

Comment:
can we pass this XML to the workspace change callback? It's already computed and it's very likely the consumer will have to serialize the blockly code anyway

Reply:
The challenge with doing that is that the XML is computed in a debounced callback, whereas the onWorkspaceChange is called immediately as soon as the workspace changes.  But now that this is a hook, one thing we could do is change the return signature to return `{ workspace, xml }` so that the caller could always have the latest XML (lagged slightly because of debouncing).